src/server/list_family.cc

```diff
@@ -195,19 +195,37 @@ OpResult<string> OpMoveSingleShard(const OpArgs& op_args, string_view src, strin
     return src_res.status();
 
   auto src_it = src_res->it;
-  quicklist* src_ql = GetQL(src_it->second);
+  quicklist* src_ql = nullptr;
+  QList* srcql_v2 = nullptr;
+  quicklist* dest_ql = nullptr;
+  QList* destql_v2 = nullptr;
+  string val;
+  size_t prev_len = 0;
 
-  if (src == dest) {  // simple case.
-    string val = ListPop(src_dir, src_ql);
+  if (src_it->second.Encoding() == OBJ_ENCODING_QUICKLIST) {
+    src_ql = GetQL(src_it->second);
+    prev_len = quicklistCount(src_ql);
+  } else {
+    DCHECK_EQ(src_it->second.Encoding(), kEncodingQL2);
+    srcql_v2 = GetQLV2(src_it->second);
+    prev_len = srcql_v2->Size();
+  }
 
-    int pos = (dest_dir == ListDir::LEFT) ? QUICKLIST_HEAD : QUICKLIST_TAIL;
-    quicklistPush(src_ql, val.data(), val.size(), pos);
+  if (src == dest) {  // simple case.
+    if (src_ql) {
+      val = ListPop(src_dir, src_ql);
+      int pos = (dest_dir == ListDir::LEFT) ? QUICKLIST_HEAD : QUICKLIST_TAIL;
+      quicklistPush(src_ql, val.data(), val.size(), pos);
+    } else {
+      val = srcql_v2->Pop(src_dir == ListDir::LEFT ? QList::HEAD : QList::TAIL);
```

Comment:
Could you create a function to make ToPosition(ListDir dir) {return  ListDir::LEFT ? QList::HEAD : QList::TAIL}